src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,340 @@
+        {
+            throw new NotSupportedException();
+        }
+
+        #endregion
+
+        #region Private methods
+
+        /// <summary>
+        /// Advances the buffer read index, and the packet data left count, by the specified length.
+        /// </summary>
+        /// <param name="length">The length to advance the packet by.</param>
+        private void AdvanceBufferOnRead(int length)
+        {
+            _readIndex += length;
+            _packetDataLeft -= length;
+        }
+
+        /// <summary>
+        /// Computes the minimum byte count available for copying into the buffer.
+        /// </summary>
+        /// <param name="maxByteCountExpected">The maximum bytes count expected by the caller.</param>
+        /// <returns></returns>
+        private int MinDataAvailableBeforeRead(int maxByteCountExpected)
+        {
+            // We can only read the minimum of what is left in the packet,
+            // what is left in the buffer, and what we need to fill
+            // If we have the max Byte Count available, then we read it
+            // else we will read either the data in packet, or the 
+            // data in buffer, whichever is smaller.
+            // If the data spans multiple packets, then the caller will go ahead and post a network read.
+            return Math.Min(Math.Min(_packetDataLeft, _readBufferDataEnd - _readIndex), maxByteCountExpected);
+        }
+
+        /// <summary>
+        /// Prepares the Read buffer with more data. 
+        /// This method is called, when the data from existing buffer is completely read, or there 
+        /// is less data available in the buffer, than what is needed to complete the read call.
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async ValueTask PrepareBufferAsync(bool isAsync, CancellationToken ct)
+        {
+            // We have read all the data from the packet as stated in the header, this means that we have to 
+            // process the next packet header.
+            if (_packetDataLeft == 0 && _readBufferDataEnd > _readIndex)
+            {
+                await ProcessHeaderAsync(isAsync, ct).ConfigureAwait(false);
+            }
+
+            // There is no data left in the buffer.
+            if (_readIndex == _readBufferDataEnd)
+            {
+                // If we have left over data indicated in the packet header, then we simply need to get data from the network.
+                if (_packetDataLeft > 0)
+                {
+                    _readBufferDataEnd = isAsync ?
+                        await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                        _underlyingStream.Read(_readBuffer);
+                    _readIndex = 0;
+                }
+                // 1.2. There is no data left as indicated by packet header and the buffer is empty.
```

Comment:
If we have 1.2 and 1.3, where's 1.1? 😃

Reply:
First my brain was structured, and came up with a numbering scheme :D 
Then it got lost in the split packet madness, and the structure was lost. 

I will fix this.

Reply:
I've definitely done the same thing before